src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,12 +394,156 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+            IList<KeySegment> keySegments = odataPath.Segments.OfType<KeySegment>().ToList();
+            if (keySegments == null)
+            {
+                return null;
+            }
+
+            foreach (KeySegment keySegment in keySegments)
+            {
+                resource.Properties = keySegment.Keys.Select(k => new ODataProperty
+                {
+                    Name = k.Key,
+                    Value = k.Value
+                });
+            }
+
+            if (refLink.EntityReferenceLink.InstanceAnnotations != null)
+            {
+                foreach (ODataInstanceAnnotation instanceAnnotation in refLink.EntityReferenceLink.InstanceAnnotations)
+                {
+                    resource.InstanceAnnotations.Add(instanceAnnotation);
+                };
+            }
+            
+            return new ODataResourceWrapper(resource);
+        }
+
+        private ODataResourceWrapper UpdateResourceWrapper(IEdmTypeReference edmPropertyType, Uri resourceId, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            Uri serviceRootUri = null;
+            string serviceRoot = "";
+            DefaultODataPathHandler pathHandler = (DefaultODataPathHandler)readContext.InternalRequest.PathHandler;
+            if (!resourceId.IsAbsoluteUri)
+            {
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+                serviceRoot = readContext.InternalUrlHelper.CreateODataLink(
+                    readContext.InternalRequest.Context.RouteName,
+                    readContext.InternalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+
+                serviceRootUri = new Uri(serviceRoot, UriKind.Absolute);
+            }
+
+            var request = readContext.Request;
+            IEdmModel model = readContext.Model;
+
+
+            ODataPath odataPath = pathHandler.Parse(serviceRoot, resourceId.ToString(), readContext.InternalRequest.RequestContainer);
+            IList<KeySegment> keySegments = odataPath.Segments.OfType<KeySegment>().ToList();
+            if (keySegments == null)
+            {
+                return null;
+            }
+
+            foreach (KeySegment keySegment in keySegments)
+            {
+                resource.Properties = keySegment.Keys.Select(k => new ODataProperty
+                {
+                    Name = k.Key,
+                    Value = k.Value
+                });
+            }
+
+            return new ODataResourceWrapper(resource);
+        }
+
+        private ODataResourceSetWrapper UpdateResourceSetWrapper(IEdmCollectionTypeReference edmPropertyType,
```

Comment:
This method is almost a mirror of `CreateResourceSetWrapper`. The key difference being a single argument plus the loop within the body that calls either `CreateResourceWrapper` or `UpdateResourceWrapper` in the different methods. You could substitute the two methods with a single one that takes an `Action` argument in place of `IList<Uri>` and `IList<ODataEntityReferenceLinkWrapper> refLinks` that differentiates them. The logic in the loop would go into the action.